### PR TITLE
Realtime effects menu grouping

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #3866

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
There are link failues on Windows to fix before merge, and there are conflicts in PluginMenus.cpp to resolve in rebasing

File: src/ListNavigationEnabled.h
Comment (Paul-Licameli) [inline src/ListNavigationEnabled.h:22]: If I wrote this, I would do the extra work to define a .cpp file, so that (1) at least one .cpp file in the system includes the header file before any other, guaranteeing that the header is self-sufficient for compilation; and (2) making out-of-line functions doing most of the work that doesn't depend on the template type parameter, leaving only small template functions in the header to be inlined.

See for instance Observer, which separates an out-of-line, "type erased" part.

I don't insist that you follow the suggestion for this PR.

Comment (Paul-Licameli) [reply src/ListNavigationEnabled.h:22]: It would also make sense to put this file in src/wrappers.  The resembles `wxPanelWrapper` for instance.  But I don't insist.  Library extractions might soon upset the folder organization anyway.  Although a library for src/wrappers would definitely be a non-toolkit-neutral part of the program until we figure out better abstractions.

File: src/CMakeLists.txt
Comment (Paul-Licameli) [inline src/CMakeLists.txt:174]: Ditto, suggestion about folder placement.  I don't insist.

File: src/CMakeLists.txt
Comment (Paul-Licameli) [inline src/CMakeLists.txt:190]: And this file.

But whatever the folders they are in,  it is good to see RealtimeEffectPanel shrink to its essentials.

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:242]: This comment got lost in the cut and paste.

Comment (Paul-Licameli) [reply src/RealtimeEffectPanel.cpp:242]: And if restored, it should also be more complete, describing what the event handlers accomplish.

File: src/menus/MenuHelper.cpp
Comment (Paul-Licameli) [inline src/menus/MenuHelper.cpp:7]: Do all these #include directives compile on all platforms?  I thought ../ would be needed in some of them.

File: src/CMakeLists.txt
Comment (Paul-Licameli) [inline src/CMakeLists.txt:773]: Maybe a more suggestive name?  PluginMenuBuilder?

File: libraries/lib-module-manager/PluginManager.cpp
Comment (Paul-Licameli) [inline libraries/lib-module-manager/PluginManager.cpp:528]: Not reachable with default Experimental.cmake but it's good you noticed it

Comment (Paul-Licameli) [reply libraries/lib-module-manager/PluginManager.cpp:528]: I see one use of Save() not followed by the publication, and that is in `PluginManager::InitializePlugins()`.  Is that right or wrong?

Comment (Paul-Licameli) [reply libraries/lib-module-manager/PluginManager.cpp:528]: Not wrong, I see, because the only subcriber is RealtimeEffectListWindow which cannot possibly be constructed earlier than that initialization.

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:746]: "its" not it's

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:905]: The constructor makes subscriptions.

`ResetTrack()` destroys both subscriptions.

`SetTrack()` restores the old subscription to non-null, but not the new one.

Is that right or wrong?

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:60]: I am glad to see you figured out the usage of MenuVisitor!  I hope it did not seem unnecessarily difficult.  If you gained some more understanding of Registry.h and Registry.cpp that is good -- they need some re-work which will need careful review by someone.

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:785]: This repeated lambda might be a local variable defined once

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:803]: I checked the function again to see what these two {}s are -- and I discovered, `realFlags` is passed around among many functions in MenuHelper.cpp but never used!

I don't know where in source code history it became unused -- but you have an opportunity to simplify with one more commit.

**Comment by Paul-Licameli** [comment]
I have finished a round of review.

The Windows builds are still failing, so you must fix that.

The question above about `SetTrack` and `ResetTrack` are the only unresolved doubt I have.

Otherwise all else is just suggestion and so this is *almost* approved.

**Comment by Paul-Licameli** [comment]
The windows build failure is because of a duplicate symbol:

```
wxmsw31u_core.lib(wxmsw313u_core_vc_custom.dll) : error LNK2005: "public: virtual __thiscall wxNavigationEnabled<class wxWindow>::~wxNavigationEnabled<class wxWindow>(void)" (??1?$wxNavigationEnabled@VwxWindow@@@@UAE@XZ) already defined in ListNavigationPanel.obj
```

... which I think you can fix just with one more `extern template` declaration at the second commit.

File: src/menus/MenuHelper.cpp
Comment (vsverchinsky) [reply src/menus/MenuHelper.cpp:7]: I believe so, yes. If I remember correctly `src` folder is added to include directories somewhere in cmake

File: src/CMakeLists.txt
Comment (vsverchinsky) [reply src/CMakeLists.txt:773]: It used to provide macros menu helpers too. And it doesn't build menus

File: src/RealtimeEffectPanel.cpp
Comment (vsverchinsky) [reply src/RealtimeEffectPanel.cpp:905]: `mPluginsChangedSubscription.Reset()` shouldn't be here, thanks!

File: src/RealtimeEffectPanel.cpp
Comment (vsverchinsky) [reply src/RealtimeEffectPanel.cpp:60]: Had to spend some time to figure out how it works, yes :)

File: src/RealtimeEffectPanel.cpp
Comment (vsverchinsky) [reply src/RealtimeEffectPanel.cpp:785]: Yep

File: src/RealtimeEffectPanel.cpp
Comment (Paul-Licameli) [inline src/RealtimeEffectPanel.cpp:786]: You made the suggested simplification to remove duplication.

Amusingly:  you could even simplify this more to:

```
      static auto realtimeEffectPredicate = &PluginDescriptor::IsEffectRealtime;
```

The implicit constructor of `std::function` used when you call `MenuHelper::PopulateEffectsMenu` works even when the template parameter is of type pointer-to-memner-function.

**Review by Paul-Licameli — APPROVED** [review]

**Comment by vsverchinsky** [comment]
Branch was rebased and few new commits were added, starting from [Effects preferences strings change](https://github.com/audacity/audacity/pull/4171/commits/537cecaaeac188f2d514e73ecd12394d9dde180e)

File: src/menus/PluginMenus.cpp
Comment (Paul-Licameli) [inline src/menus/PluginMenus.cpp]: The definition of `IsRealtimeNotActiveFlag` is in this file.  It becomes unused, so may be deleted too.

File: src/ThemedWrappers.h
Comment (Paul-Licameli) [inline src/ThemedWrappers.h:33]: Maybe this is more elegant:

```
template<class WindowBase, class... OtherBases>
class ThemedWindowWrapper final :
   public WindowBase, public OtherBases...,
```

So you can simply use `PrefsListener` where needed as another template argument, and not mention that class name here.

File: src/ThemedWrappers.h
Comment (Paul-Licameli) [inline src/ThemedWrappers.h:72]: Maybe you could use composition instead of inheritance:  a member of a type inheriting PrefsListener.  Then you would not need to play with type traits to pass through to the base.  Instead UpdatePrefs is directly invoked directly on this member and on the base class to do different unrelated things that can happen in either sequence.

File: src/menus/MenuHelper.cpp
Comment (Paul-Licameli) [inline src/menus/MenuHelper.cpp:148]: This variable is not longer used, so remove it too

File: src/prefs/EffectsPrefs.cpp
Comment (Paul-Licameli) [inline src/prefs/EffectsPrefs.cpp:132]: A wise comment.  I like it that you are aware of dependencies you add.

But I checked that this does not add a dependency cycle.

True, it may still be an undesirable dependency, but we can figure out some trick to break it if we need that later.

File: src/ThemedWrappers.h
Comment (Paul-Licameli) [reply src/ThemedWrappers.h:72]: ... or maybe, even without composition, all you need is:
```
if constexpr(std::is_base_of_v<PrefsListener, WindowBase>)
   WindowBase::UpdatePrefs();
```

**Review by Paul-Licameli — APPROVED** [review]
I caught up at last with the new version of `AddGroupedEffectMenuItems` and I'm finished with this review.

**Review by Paul-Licameli — APPROVED** [review]

**Comment by vsverchinsky** [comment]
> I note that the default setttings for the Effect Menu and the Realtime effects are different. Is this intentional, by design, or are they supposed to be the same by default?

Yes, they are different by design

**Pull request merged**
